oak-search/src/main/java/org/apache/jackrabbit/oak/plugins/index/search/spi/query/FulltextIndexTracker.java

```diff
@@ -46,18 +47,16 @@
 import static com.google.common.base.Predicates.in;
 import static com.google.common.base.Predicates.not;
 import static com.google.common.base.Predicates.notNull;
-import static com.google.common.collect.Lists.newArrayListWithCapacity;
 import static com.google.common.collect.Maps.filterKeys;
 import static com.google.common.collect.Maps.filterValues;
-import static com.google.common.collect.Maps.newHashMap;
 import static java.util.Collections.emptyMap;
 import static org.apache.jackrabbit.oak.plugins.index.search.IndexDefinition.INDEX_DEFINITION_NODE;
 import static org.apache.jackrabbit.oak.plugins.index.search.IndexDefinition.STATUS_NODE;
 import static org.apache.jackrabbit.oak.plugins.memory.EmptyNodeState.EMPTY_NODE;
 
-public abstract class FulltextIndexTracker {
+public abstract class FulltextIndexTracker<N extends IndexNode, I extends IndexNodeManager<N>> {
```

Comment:
I (think) I understand this, but I have to admit it makes it much harder to read... I wonder if there is an alternative...

Reply:
it would have been a lot nicer to have something like

```java
public abstract class FulltextIndexTracker<I extends IndexNodeManager<N>> {
```

with `N extends IndexNode` inferred from the `IndexNodeManager` but Java does not seem very smart in this case.

An alternative would be to change the signature of `acquireIndexNode` to return an `IndexNodeManager` instead of an `IndexNode`